File: openclaw-plugin/src/attachments.ts
import { realpath } from "node:fs/promises";
import { basename, resolve, sep } from "node:path";

import type { PumpUp, PumpUpClient } from "../../../sdks/typescript/index.js";

/// One agent-supplied attachment: a workspace-relative file path + display label.
export type AttachmentInput = { path: string; label: string };

/// Upload workspace-relative files to Pump Up and return task attachment refs. Paths are confined to
/// the agent workspace (data-egress guard); a failed upload propagates for structured error mapping.
export async function uploadAttachments(
  client: PumpUpClient,
  workspaceDir: string | undefined,
  inputs: AttachmentInput[] | undefined,
): Promise<PumpUp.Attachment[]> {
  if (!inputs || inputs.length === 0) {
    return [];
  }
  if (!workspaceDir) {
    throw new Error("pumpup: attachments require an agent workspace directory");
  }
  // realpath both sides so a symlink inside the workspace can't point the upload at a file outside it.
  const root = await realpath(resolve(workspaceDir));
  const refs: PumpUp.Attachment[] = [];
  for (const input of inputs) {
    const abs = await realpath(resolve(root, input.path));
    if (abs !== root && !abs.startsWith(root + sep)) {
      throw new Error(`pumpup: attachment path escapes the workspace: ${input.path}`);
    }
    const uploaded = await client.uploads.upload({ file: { path: abs, filename: basename(abs) } });
    refs.push({ uploadId: uploaded.id, label: input.label });
  }
  return refs;
}

File: hermes-plugin/pumpup_hermes/attachments.py
"""Upload workspace-relative files to Pump Up, confined to the agent's working directory.

Symlinks are resolved (realpath) on both the root and each path so a link inside the workspace
can't point the upload at a file outside it (a data-egress guard). Hermes file tools operate
relative to TERMINAL_CWD / cwd; attachments confine to the same root (override PUMPUP_WORKSPACE_DIR).
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Any, Optional

from pumpup import AsyncPumpUp, Attachment


def _workspace_root() -> Path:
    """The directory agent-supplied attachment paths are resolved against and confined to."""
    raw = os.getenv("PUMPUP_WORKSPACE_DIR") or os.getenv("TERMINAL_CWD") or os.getcwd()
    return Path(os.path.realpath(raw))


async def upload_attachments(client: AsyncPumpUp, inputs: Optional[list[dict[str, Any]]]) -> list[Attachment]:
    """Upload each {path, label} and return task attachment refs; raises if a path escapes the workspace."""
    if not inputs:
        return []
    root = _workspace_root()
    refs: list[Attachment] = []
    for item in inputs:
        path = item["path"]
        abs_path = Path(os.path.realpath(root / path))
        if abs_path != root and root not in abs_path.parents:
            raise ValueError(f"pumpup: attachment path escapes the workspace: {path}")
        uploaded = await client.uploads.upload(file=(abs_path.name, abs_path.read_bytes()))
        refs.append(Attachment(upload_id=uploaded.id, label=item["label"]))
    return refs
